src/ghostwriter/models/v1/mapping.py

```diff
@@ -0,0 +1,180 @@
+"""Definition of the mapping from a top-level route to its substituted form,
+as well as any hooks that should be called before passing the input route
+to be rewritten.
+"""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from string import Template
+from typing import Annotated
+
+from pydantic import AfterValidator, BaseModel, BeforeValidator, Field
+
+from ... import hooks
+from ...exceptions import (
+    HookError,
+    HookNotFoundError,
+    MatchNotFoundError,
+    ResolutionError,
+)
+from ..substitution import Parameters
+
+
+def canonicalize_source_route(v: str) -> str:
+    """Force source route to have one leading and one trailing slash."""
+    return f"/{v.strip('/')}/"
+
+
+def resolve_hooks(
+    v: list[str] | list[Callable[[Parameters], Awaitable[None]]],
+) -> list[Callable[[Parameters], Awaitable[None]]]:
+    """Hooks will be listed as strings in the config file.  This loads
+    the corresponding functions.
+    """
+    retval: list[Callable[[Parameters], Awaitable[None]]] = []
+    for hook in v:
+        if callable(hook):
+            retval.append(hook)
+            continue
+        if not isinstance(hook, str):
+            raise HookNotFoundError(f"Hook {hook} could not be loaded")
+        if hook.startswith("ghostwriter.hooks."):
+            hookname = hook[len("ghostwriter.hooks.") :]
+        else:
+            hookname = hook
+        try:
+            obj = getattr(hooks, hookname)
+        except AttributeError as exc:
+            raise HookNotFoundError(
+                f"Hookname {hook} could not be loaded: {exc}"
+            ) from exc
+        retval.append(obj)
+    return retval
+
+
+class MapRule(BaseModel):
```

Comment:
Suggested name change:
`MapRule` -> `RouteMapping`

This is nitpicky but I think the RouteMapping slightly matches the model a bit more